File: src/Core/DeltaCodec.fs
namespace Zeta.Core

open System.Collections.Generic


/// **Pluggable serialization seam for the durable delta log.** Encodes a
/// `ZSet<'K>` delta to bytes and back. Lets `DeltaLog`/`RecoverableSpine` stay
/// format-agnostic — land on `Checkpoint` today, swap to the byte-verified
/// canonical CBOR / YAML codecs later, without touching log/recovery logic (the
/// "stable contract, swap format" path; companion doc §9). MUST be a lossless
/// round-trip: `Decode (Encode z) = z`.
type IDeltaCodec<'K when 'K : comparison> = IDeltaCodec<'K, ZSet<'K>>


/// `ZSet` ↔ `DynamicValue` mapping — the bridge that lets a Z-set ride ALL our
/// byte-verified serializers (CBOR / YAML / JSON / Arrow). A `ZSet<'K>` becomes a
/// `DynamicValue.Array` of `[keyDv; Int weight]` pairs, given a per-key codec
/// (keys are `'K`, so the caller supplies key ↔ `DynamicValue`).
[<CompilationRepresentation(CompilationRepresentationFlags.ModuleSuffix)>]
module ZSetDynamic =

    let toDynamicValue (keyEnc: 'K -> DynamicValue) (z: ZSet<'K>) : DynamicValue =
        DynamicValue.Array
            [ for e in z -> DynamicValue.Array [ keyEnc e.Key; DynamicValue.Int e.Weight ] ]

    let ofDynamicValue (keyDec: DynamicValue -> 'K) (dv: DynamicValue) : ZSet<'K> =
        match dv with
        | DynamicValue.Array pairs ->
            pairs
            |> List.map (fun p ->
                match p with
                | DynamicValue.Array [ k; DynamicValue.Int w ] -> keyDec k, w
                | other ->
                    invalidArg (nameof dv) $"ZSetDynamic: expected [key, Int weight] pair, got {other}")
            |> ZSet.ofSeq
        | other -> invalidArg (nameof dv) $"ZSetDynamic: expected Array, got {other}"


/// Default codec: wraps the existing `Checkpoint` serializer (works for any `'K`
/// `Checkpoint` supports). The land-it-today baseline behind the seam — the disk
/// log and benchmark start here, then swap in CBOR/YAML once measured.
[<Sealed>]
type CheckpointDeltaCodec<'K when 'K : comparison>() =
    interface IDeltaCodec<'K> with
        member _.Encode z = Checkpoint.toBytes z
        member _.Decode bytes = Checkpoint.ofBytes<'K> bytes


/// Byte-verified canonical **CBOR** codec (the binary hot/perf tier). Maps
/// `ZSet` ↔ `DynamicValue` via the supplied key codec, then rides DynamicValue's
/// golden-vector-locked CBOR (RFC 8949, complete 8/8 shapes). For a self-produced
/// log the canonicality re-encode in `fromCanonicalCbor` is redundant work — a
/// `trustCanonical` fast path is a documented follow-up (Naledi, companion §9).
[<Sealed>]
type CborDeltaCodec<'K when 'K : comparison>
    (keyEnc: 'K -> DynamicValue, keyDec: DynamicValue -> 'K) =
    interface IDeltaCodec<'K> with
        member _.Encode z =
            ZSetDynamic.toDynamicValue keyEnc z |> DynamicValue.toCanonicalCborOk
        member _.Decode bytes =
            match DynamicValue.fromCanonicalCbor bytes with
            | Ok dv -> ZSetDynamic.ofDynamicValue keyDec dv
            | Error e -> invalidArg (nameof bytes) $"CborDeltaCodec.Decode: non-decodable CBOR: {e}"


/// `DeltaLogEntry` ↔ `DynamicValue` mapping — the canonical **Log-entry envelope**, the
/// `Log` noun's path to the 4-language proven bar (workitem 081KTGD5JMD). A whole entry
/// `{ Seq; Delta; Captured }` becomes a `DynamicValue.Object` with the keys
/// `captured` / `delta` / `seq` (ordinal order), so the entry rides DynamicValue's
/// golden-vector-locked canonical CBOR — the Log entry INHERITS the existing 4-lang
/// byte-lock rather than inventing a new canonical encoding (it adds no new noun: an
/// entry is just a `DynamicValue`). `Captured` keys are **ordinal-sorted**
/// (culture-invariant; 081KT07NV0008QG0R001YDB73K) for cross-language + DST determinism — the prior
/// `System.Text.Json` `Dictionary` encoding in `GitDeltaLog`/`DiskDeltaLog` is NOT
/// canonical and is the gap this closes. Lossless round-trip: `ofDynamicValue (toDynamicValue e) = e`.
[<CompilationRepresentation(CompilationRepresentationFlags.ModuleSuffix)>]
module DeltaLogEntryDynamic =

    let toDynamicValue (keyEnc: 'K -> DynamicValue) (entry: DeltaLogEntry<'K>) : DynamicValue =
        let captured =
            entry.Captured
            |> Seq.map (fun kv -> kv.Key, kv.Value)
            |> Seq.toList
            |> List.sortWith (fun (a, _) (b, _) -> System.String.CompareOrdinal(a, b))
            |> List.map (fun (k, v) -> k, DynamicValue.String v)
        DynamicValue.Object
            [ "captured", DynamicValue.Object captured
              "delta", ZSetDynamic.toDynamicValue keyEnc entry.Delta
              "seq", DynamicValue.Int entry.Seq ]

    let ofDynamicValue (keyDec: DynamicValue -> 'K) (dv: DynamicValue) : DeltaLogEntry<'K> =
        match dv with
        | DynamicValue.Object fields ->
            let find name =
                match List.tryFind (fun (k, _) -> k = name) fields with
                | Some(_, v) -> v
                | None -> invalidArg (nameof dv) $"DeltaLogEntryDynamic: missing field '{name}'"
            let seq =
                match find "seq" with
                | DynamicValue.Int s -> s
                | o -> invalidArg (nameof dv) $"DeltaLogEntryDynamic: 'seq' not Int: {o}"
            let delta = ZSetDynamic.ofDynamicValue keyDec (find "delta")
            let captured =
                match find "captured" with
                | DynamicValue.Object kvs ->
                    kvs
                    |> List.map (fun (k, v) ->
                        match v with
                        | DynamicValue.String s -> k, s
                        | o -> invalidArg (nameof dv) $"DeltaLogEntryDynamic: captured['{k}'] not String: {o}")
                    |> Map.ofList
                | o -> invalidArg (nameof dv) $"DeltaLogEntryDynamic: 'captured' not Object: {o}"
            DeltaLogEntry<'K>(seq, delta, captured)
        | other -> invalidArg (nameof dv) $"DeltaLogEntryDynamic: expected Object, got {other}"


/// Format-parameterized canonical codec for a whole `DeltaLogEntry` (Seq + Delta + Captured), the
/// 4-language treaty target for the `Log` noun. The format-INDEPENDENT core is the
/// `DeltaLogEntryDynamic` mapping; each concrete format rides DynamicValue's already-byte-locked
/// per-format serializer, so the bytes/text are reproducible across F#/C#/Rust/TS.
///
/// Per-stream/table format choice (Aaron 2026-06-07): **git check-ins default to YAML** (diffable,
/// human-readable history) and **filesystem defaults to CBOR** (speed); all formats are optionally
/// selectable per stream/table at the backend layer. CBOR/JSON/XML round-trip exists today; **YAML is
/// a prerequisite gap** (DynamicValue has no `toYaml`/`fromYaml` yet — tracked for the git-default path).
/// These functions are the canonical replacement for the per-backend `System.Text.Json` framing in
/// `GitDeltaLog`/`DiskDeltaLog`.
[<CompilationRepresentation(CompilationRepresentationFlags.ModuleSuffix)>]
module DeltaLogEntryCodec =

    /// Canonical CBOR (the filesystem default — speed; complete 8/8 shapes, full round-trip).
    let encodeCbor (keyEnc: 'K -> DynamicValue) (entry: DeltaLogEntry<'K>) : byte[] =
        DeltaLogEntryDynamic.toDynamicValue keyEnc entry |> DynamicValue.toCanonicalCborOk

    let decodeCbor (keyDec: DynamicValue -> 'K) (bytes: byte[]) : DeltaLogEntry<'K> =
        match DynamicValue.fromCanonicalCbor bytes with
        | Ok dv -> DeltaLogEntryDynamic.ofDynamicValue keyDec dv
        | Error e -> invalidArg (nameof bytes) $"DeltaLogEntryCodec.decodeCbor: non-decodable CBOR: {e}"

    /// Canonical JSON (text — a diff-friendly option until the YAML git-default serializer lands).
    let encodeJson (keyEnc: 'K -> DynamicValue) (entry: DeltaLogEntry<'K>) : Result<string, EncodeError> =
        DeltaLogEntryDynamic.toDynamicValue keyEnc entry |> DynamicValue.toCanonicalJson

    let decodeJson (keyDec: DynamicValue -> 'K) (json: string) : DeltaLogEntry<'K> =
        match DynamicValue.fromCanonicalJson json with
        | Ok dv -> DeltaLogEntryDynamic.ofDynamicValue keyDec dv
        | Error e -> invalidArg (nameof json) $"DeltaLogEntryCodec.decodeJson: non-decodable JSON: {e}"


/// **Whole-entry serialization seam** — encodes a full `DeltaLogEntry` (Seq + Delta + Captured) to
/// canonical bytes and back. The migration target for the durable backends (`GitDeltaLog` /
/// `DiskDeltaLog`): serialize the WHOLE entry through this — the proven, 4-language byte-locked
/// `DeltaLogEntryCodec` format (golden seed `src/Core.TypeScript/delta-log-entry/golden-vectors.json`) —
/// replacing the per-backend `System.Text.Json` framing of the `Captured` map. Distinct from
/// `IDeltaCodec` (which encodes only the ZSet delta): an entry is `(Seq, Delta, Captured)`, and the whole
/// entry is the cross-language treaty unit. MUST be a lossless round-trip: `Decode (Encode e) = e`.
type IEntryCodec<'K when 'K : comparison> = IEntryCodec<'K, ZSet<'K>>


/// Canonical **CBOR** whole-entry codec — rides `DeltaLogEntryCodec` (the DynamicValue-locked canonical
/// CBOR). The backend migration target: byte-identical across F#/C#/Rust/TS, so the on-disk / in-git
/// bytes ARE the cross-language treaty. The `Seq` rides inside the entry, so `Decode` needs no external
/// sequence argument.
[<Sealed>]
type CborEntryCodec<'K when 'K : comparison>
    (keyEnc: 'K -> DynamicValue, keyDec: DynamicValue -> 'K) =
    interface IEntryCodec<'K> with
        member _.Encode entry = DeltaLogEntryCodec.encodeCbor keyEnc entry
        member _.Decode bytes = DeltaLogEntryCodec.decodeCbor keyDec bytes

File: src/Core.CSharp/DeltaCodec.cs
#pragma warning disable MA0048

using System;
using System.Collections.Generic;
using System.Collections.Immutable;
using System.Linq;

namespace Zeta.Core.CSharp;

/// <summary>
/// ZSet ↔ DynamicValue mapping — the bridge that lets a Z-set ride all our
/// byte-verified serializers. A ZSet&lt;K&gt; becomes a DynamicValue.Array
/// of [keyDv; Int weight] pairs.
/// </summary>
public static class ZSetDynamic
{
    /// <summary>
    /// Encodes a ZSet to a DynamicValue.
    /// </summary>
    /// <typeparam name="TKey">The key type.</typeparam>
    /// <param name="keyEnc">The key encoder function.</param>
    /// <param name="z">The ZSet to encode.</param>
    /// <returns>A DynamicValue representation of the ZSet.</returns>
    public static DynamicValue ToDynamicValue<TKey>(Func<TKey, DynamicValue> keyEnc, ZSet<TKey> z)
    {
        ArgumentNullException.ThrowIfNull(keyEnc);
        ArgumentNullException.ThrowIfNull(z);
        var entries = z.ToImmutableArray();
        var builder = ImmutableArray.CreateBuilder<DynamicValue>(entries.Length);
        foreach (var e in entries)
        {
            builder.Add(new DynamicValue.Array(ImmutableArray.Create(
                keyEnc(e.Key),
                new DynamicValue.Int(e.Weight)
            )));
        }
        return new DynamicValue.Array(builder.ToImmutable());
    }

    /// <summary>
    /// Decodes a ZSet from a DynamicValue.
    /// </summary>
    /// <typeparam name="TKey">The key type.</typeparam>
    /// <param name="keyDec">The key decoder function.</param>
    /// <param name="dv">The DynamicValue to decode.</param>
    /// <param name="comparer">Optional key comparer.</param>
    /// <param name="collation">The named collation to use.</param>
    /// <returns>The decoded ZSet.</returns>
    public static ZSet<TKey> OfDynamicValue<TKey>(Func<DynamicValue, TKey> keyDec, DynamicValue dv, IComparer<TKey>? comparer = null, string collation = Collation.DefaultName)
    {
        ArgumentNullException.ThrowIfNull(keyDec);
        ArgumentNullException.ThrowIfNull(dv);
        if (dv is not DynamicValue.Array arrayDv)
        {
            throw new ArgumentException($"ZSetDynamic: expected Array, got {dv.Type}", nameof(dv));
        }

        var list = new List<(TKey Key, long Weight)>(arrayDv.Items.Length);
        foreach (var item in arrayDv.Items)
        {
            if (item is not DynamicValue.Array pair || pair.Items.Length != 2)
            {
                throw new ArgumentException($"ZSetDynamic: expected [key, Int weight] pair, got {item}", nameof(dv));
            }
            if (pair.Items[1] is not DynamicValue.Int w)
            {
                throw new ArgumentException($"ZSetDynamic: expected [key, Int weight] pair, got {item}", nameof(dv));
            }
            list.Add((keyDec(pair.Items[0]), w.Value));
        }

        var cmp = comparer ?? Collation.ForKey<TKey>(collation);
        return ZSet.OfEntries(list, cmp, collation);
    }
}

/// <summary>
/// Byte-verified canonical CBOR codec. Maps ZSet ↔ DynamicValue via the supplied key codec,
/// then rides DynamicValue's golden-vector-locked CBOR.
/// </summary>
/// <typeparam name="TKey">The key type.</typeparam>
public sealed class CborDeltaCodec<TKey> : IDeltaCodec<TKey, ZSet<TKey>>
{
    private readonly Func<TKey, DynamicValue> _keyEnc;
    private readonly Func<DynamicValue, TKey> _keyDec;
    private readonly IComparer<TKey>? _comparer;
    private readonly string _collation;

    /// <summary>
    /// Initializes a new instance of the <see cref="CborDeltaCodec{TKey}"/> class.
    /// </summary>
    /// <param name="keyEnc">The key encoder function.</param>
    /// <param name="keyDec">The key decoder function.</param>
    /// <param name="comparer">Optional key comparer.</param>
    /// <param name="collation">Collation name.</param>
    public CborDeltaCodec(Func<TKey, DynamicValue> keyEnc, Func<DynamicValue, TKey> keyDec, IComparer<TKey>? comparer = null, string collation = Collation.DefaultName)
    {
        _keyEnc = keyEnc ?? throw new ArgumentNullException(nameof(keyEnc));
        _keyDec = keyDec ?? throw new ArgumentNullException(nameof(keyDec));
        _comparer = comparer;
        _collation = collation;
    }

    /// <summary>
    /// Encodes the ZSet to canonical CBOR bytes.
    /// </summary>
    /// <param name="state">The ZSet state to encode.</param>
    /// <returns>The encoded CBOR bytes.</returns>
    public byte[] Encode(ZSet<TKey> state)
    {
        var dv = ZSetDynamic.ToDynamicValue(_keyEnc, state);
        return DynamicValues.ToCanonicalCborOk(dv);
    }

    /// <summary>
    /// Decodes the ZSet from canonical CBOR bytes.
    /// </summary>
    /// <param name="bytes">The CBOR bytes to decode.</param>
    /// <returns>The decoded ZSet state.</returns>
    public ZSet<TKey> Decode(byte[] bytes)
    {
        var res = DynamicValues.FromCanonicalCbor(bytes);
        if (res is Result<DynamicValue, DecodeError>.Ok ok)
        {
            return ZSetDynamic.OfDynamicValue(_keyDec, ok.Value, _comparer, _collation);
        }
        else
        {
            var err = (Result<DynamicValue, DecodeError>.Err)res;
            throw new ArgumentException($"CborDeltaCodec.Decode: non-decodable CBOR: {err.Error}", nameof(bytes));
        }
    }
}

/// <summary>
/// DeltaLogEntry ↔ DynamicValue mapping — the canonical Log-entry envelope.
/// A whole entry { Seq; Delta; Captured } becomes a DynamicValue.Object with keys
/// "captured" / "delta" / "seq" (ordinal order).
/// </summary>
public static class DeltaLogEntryDynamic
{
    /// <summary>
    /// Encodes a DeltaLogEntry to a DynamicValue.
    /// </summary>
    /// <typeparam name="TKey">The key type.</typeparam>
    /// <param name="keyEnc">The key encoder function.</param>
    /// <param name="entry">The DeltaLogEntry to encode.</param>
    /// <returns>A DynamicValue representation of the entry.</returns>
    public static DynamicValue ToDynamicValue<TKey>(Func<TKey, DynamicValue> keyEnc, DeltaLogEntry<TKey, ZSet<TKey>> entry)
    {
        ArgumentNullException.ThrowIfNull(keyEnc);
        ArgumentNullException.ThrowIfNull(entry);

        var sortedPairs = entry.Captured
            .OrderBy(kv => kv.Key, StringComparer.Ordinal)
            .Select(kv => new KeyValuePair<string, DynamicValue>(kv.Key, new DynamicValue.String(kv.Value)))
            .ToImmutableArray();

        var fields = ImmutableArray.Create(
            new KeyValuePair<string, DynamicValue>("captured", new DynamicValue.Object(sortedPairs)),
            new KeyValuePair<string, DynamicValue>("delta", ZSetDynamic.ToDynamicValue(keyEnc, entry.Delta)),
            new KeyValuePair<string, DynamicValue>("seq", new DynamicValue.Int(entry.Seq))
        );

        return new DynamicValue.Object(fields);
    }

    /// <summary>
    /// Decodes a DeltaLogEntry from a DynamicValue.
    /// </summary>
    /// <typeparam name="TKey">The key type.</typeparam>
    /// <param name="keyDec">The key decoder function.</param>
    /// <param name="dv">The DynamicValue to decode.</param>
    /// <param name="comparer">Optional key comparer.</param>
    /// <param name="collation">The collation name.</param>
    /// <returns>The decoded DeltaLogEntry.</returns>
    public static DeltaLogEntry<TKey, ZSet<TKey>> OfDynamicValue<TKey>(
        Func<DynamicValue, TKey> keyDec,
        DynamicValue dv,
        IComparer<TKey>? comparer = null,
        string collation = Collation.DefaultName)
    {
        ArgumentNullException.ThrowIfNull(keyDec);
        ArgumentNullException.ThrowIfNull(dv);

        if (dv is not DynamicValue.Object objDv)
        {
            throw new ArgumentException($"DeltaLogEntryDynamic: expected Object, got {dv.Type}", nameof(dv));
        }

        DynamicValue? FindField(string name)
        {
            foreach (var pair in objDv.Pairs)
            {
                if (string.Equals(pair.Key, name, StringComparison.Ordinal))
                {
                    return pair.Value;
                }
            }
            return null;
        }

        var seqDv = FindField("seq") ?? throw new ArgumentException("DeltaLogEntryDynamic: missing field 'seq'", nameof(dv));
        if (seqDv is not DynamicValue.Int seqInt)
        {
            throw new ArgumentException($"DeltaLogEntryDynamic: 'seq' not Int: {seqDv.Type}", nameof(dv));
        }
        long seq = seqInt.Value;

        var deltaDv = FindField("delta") ?? throw new ArgumentException("DeltaLogEntryDynamic: missing field 'delta'", nameof(dv));
        var delta = ZSetDynamic.OfDynamicValue(keyDec, deltaDv, comparer, collation);

        var capturedDv = FindField("captured") ?? throw new ArgumentException("DeltaLogEntryDynamic: missing field 'captured'", nameof(dv));
        if (capturedDv is not DynamicValue.Object capturedObj)
        {
            throw new ArgumentException($"DeltaLogEntryDynamic: 'captured' not Object: {capturedDv.Type}", nameof(dv));
        }

        var capturedDict = new Dictionary<string, string>(StringComparer.Ordinal);
        foreach (var pair in capturedObj.Pairs)
        {
            if (pair.Value is not DynamicValue.String strVal)
            {
                throw new ArgumentException($"DeltaLogEntryDynamic: captured['{pair.Key}'] not String: {pair.Value.Type}", nameof(dv));
            }
            capturedDict[pair.Key] = strVal.Value;
        }

        return new DeltaLogEntry<TKey, ZSet<TKey>>(seq, delta, capturedDict);
    }
}

/// <summary>
/// Format-parameterized canonical codec for a whole DeltaLogEntry (Seq + Delta + Captured).
/// </summary>
public static class DeltaLogEntryCodec
{
    /// <summary>
    /// Encodes the entry to CBOR bytes.
    /// </summary>
    /// <typeparam name="TKey">The key type.</typeparam>
    /// <param name="keyEnc">The key encoder function.</param>
    /// <param name="entry">The DeltaLogEntry to encode.</param>
    /// <returns>The encoded CBOR bytes.</returns>
    public static byte[] EncodeCbor<TKey>(Func<TKey, DynamicValue> keyEnc, DeltaLogEntry<TKey, ZSet<TKey>> entry)
    {
        var dv = DeltaLogEntryDynamic.ToDynamicValue(keyEnc, entry);
        return DynamicValues.ToCanonicalCborOk(dv);
    }

    /// <summary>
    /// Decodes the entry from CBOR bytes.
    /// </summary>
    /// <typeparam name="TKey">The key type.</typeparam>
    /// <param name="keyDec">The key decoder function.</param>
    /// <param name="bytes">The CBOR bytes to decode.</param>
    /// <param name="comparer">Optional key comparer.</param>
    /// <param name="collation">The collation name.</param>
    /// <returns>The decoded DeltaLogEntry.</returns>
    public static DeltaLogEntry<TKey, ZSet<TKey>> DecodeCbor<TKey>(
        Func<DynamicValue, TKey> keyDec,
        byte[] bytes,
        IComparer<TKey>? comparer = null,
        string collation = Collation.DefaultName)
    {
        var res = DynamicValues.FromCanonicalCbor(bytes);
        if (res is Result<DynamicValue, DecodeError>.Ok ok)
        {
            return DeltaLogEntryDynamic.OfDynamicValue(keyDec, ok.Value, comparer, collation);
        }
        else
        {
            var err = (Result<DynamicValue, DecodeError>.Err)res;
            throw new ArgumentException($"DeltaLogEntryCodec.DecodeCbor: non-decodable CBOR: {err.Error}", nameof(bytes));
        }
    }

    /// <summary>
    /// Encodes the entry to a JSON string.
    /// </summary>
    /// <typeparam name="TKey">The key type.</typeparam>
    /// <param name="keyEnc">The key encoder function.</param>
    /// <param name="entry">The entry to encode.</param>
    /// <returns>The encoded JSON result.</returns>
    public static Result<string, EncodeError> EncodeJson<TKey>(Func<TKey, DynamicValue> keyEnc, DeltaLogEntry<TKey, ZSet<TKey>> entry)
    {
        var dv = DeltaLogEntryDynamic.ToDynamicValue(keyEnc, entry);
        return DynamicValues.ToCanonicalJson(dv);
    }

    /// <summary>
    /// Decodes the entry from a JSON string.
    /// </summary>
    /// <typeparam name="TKey">The key type.</typeparam>
    /// <param name="keyDec">The key decoder function.</param>
    /// <param name="json">The JSON string.</param>
    /// <param name="comparer">Optional key comparer.</param>
    /// <param name="collation">The collation name.</param>
    /// <returns>The decoded DeltaLogEntry.</returns>
    public static DeltaLogEntry<TKey, ZSet<TKey>> DecodeJson<TKey>(
        Func<DynamicValue, TKey> keyDec,
        string json,
        IComparer<TKey>? comparer = null,
        string collation = Collation.DefaultName)
    {
        var res = DynamicValues.FromCanonicalJson(json);
        if (res is Result<DynamicValue, DecodeError>.Ok ok)
        {
            return DeltaLogEntryDynamic.OfDynamicValue(keyDec, ok.Value, comparer, collation);
        }
        else
        {
            var err = (Result<DynamicValue, DecodeError>.Err)res;
            throw new ArgumentException($"DeltaLogEntryCodec.DecodeJson: non-decodable JSON: {err.Error}", nameof(json));
        }
    }

    /// <summary>
    /// Encodes the entry to a YAML string.
    /// </summary>
    /// <typeparam name="TKey">The key type.</typeparam>
    /// <param name="keyEnc">The key encoder function.</param>
    /// <param name="entry">The entry to encode.</param>
    /// <returns>The encoded YAML result.</returns>
    public static Result<string, EncodeError> EncodeYaml<TKey>(Func<TKey, DynamicValue> keyEnc, DeltaLogEntry<TKey, ZSet<TKey>> entry)
    {
        var dv = DeltaLogEntryDynamic.ToDynamicValue(keyEnc, entry);
        return DynamicValues.ToYaml(dv);
    }

    /// <summary>
    /// Decodes the entry from a YAML string.
    /// </summary>
    /// <typeparam name="TKey">The key type.</typeparam>
    /// <param name="keyDec">The key decoder function.</param>
    /// <param name="yaml">The YAML string.</param>
    /// <param name="comparer">Optional key comparer.</param>
    /// <param name="collation">The collation name.</param>
    /// <returns>The decoded DeltaLogEntry.</returns>
    public static DeltaLogEntry<TKey, ZSet<TKey>> DecodeYaml<TKey>(
        Func<DynamicValue, TKey> keyDec,
        string yaml,
        IComparer<TKey>? comparer = null,
        string collation = Collation.DefaultName)
    {
        var res = DynamicValues.FromYaml(yaml);
        if (res is Result<DynamicValue, DecodeError>.Ok ok)
        {
            return DeltaLogEntryDynamic.OfDynamicValue(keyDec, ok.Value, comparer, collation);
        }
        else
        {
            var err = (Result<DynamicValue, DecodeError>.Err)res;
            throw new ArgumentException($"DeltaLogEntryCodec.DecodeYaml: non-decodable YAML: {err.Error}", nameof(yaml));
        }
    }
}

/// <summary>
/// Canonical CBOR whole-entry codec — rides DeltaLogEntryCodec.
/// </summary>
/// <typeparam name="TKey">The key type.</typeparam>
public sealed class CborEntryCodec<TKey> : IEntryCodec<TKey, ZSet<TKey>>
{
    private readonly Func<TKey, DynamicValue> _keyEnc;
    private readonly Func<DynamicValue, TKey> _keyDec;
    private readonly IComparer<TKey>? _comparer;
    private readonly string _collation;

    /// <summary>
    /// Initializes a new instance of the <see cref="CborEntryCodec{TKey}"/> class.
    /// </summary>
    /// <param name="keyEnc">The key encoder function.</param>
    /// <param name="keyDec">The key decoder function.</param>
    /// <param name="comparer">Optional key comparer.</param>
    /// <param name="collation">The collation name.</param>
    public CborEntryCodec(Func<TKey, DynamicValue> keyEnc, Func<DynamicValue, TKey> keyDec, IComparer<TKey>? comparer = null, string collation = Collation.DefaultName)
    {
        _keyEnc = keyEnc ?? throw new ArgumentNullException(nameof(keyEnc));
        _keyDec = keyDec ?? throw new ArgumentNullException(nameof(keyDec));
        _comparer = comparer;
        _collation = collation;
    }

    /// <summary>
    /// Encodes a DeltaLogEntry to bytes.
    /// </summary>
    public byte[] Encode(DeltaLogEntry<TKey, ZSet<TKey>> entry)
    {
        return DeltaLogEntryCodec.EncodeCbor(_keyEnc, entry);
    }

    /// <summary>
    /// Decodes a DeltaLogEntry from bytes.
    /// </summary>
    public DeltaLogEntry<TKey, ZSet<TKey>> Decode(byte[] bytes)
    {
        return DeltaLogEntryCodec.DecodeCbor(_keyDec, bytes, _comparer, _collation);
    }
}

/// <summary>
/// Canonical YAML whole-entry codec — rides DeltaLogEntryCodec.
/// </summary>
/// <typeparam name="TKey">The key type.</typeparam>
public sealed class YamlEntryCodec<TKey> : IEntryCodec<TKey, ZSet<TKey>>
{
    private readonly Func<TKey, DynamicValue> _keyEnc;
    private readonly Func<DynamicValue, TKey> _keyDec;
    private readonly IComparer<TKey>? _comparer;
    private readonly string _collation;

    /// <summary>
    /// Initializes a new instance of the <see cref="YamlEntryCodec{TKey}"/> class.
    /// </summary>
    /// <param name="keyEnc">The key encoder function.</param>
    /// <param name="keyDec">The key decoder function.</param>
    /// <param name="comparer">Optional key comparer.</param>
    /// <param name="collation">The collation name.</param>
    public YamlEntryCodec(Func<TKey, DynamicValue> keyEnc, Func<DynamicValue, TKey> keyDec, IComparer<TKey>? comparer = null, string collation = Collation.DefaultName)
    {
        _keyEnc = keyEnc ?? throw new ArgumentNullException(nameof(keyEnc));
        _keyDec = keyDec ?? throw new ArgumentNullException(nameof(keyDec));
        _comparer = comparer;
        _collation = collation;
    }

    /// <summary>
    /// Encodes a DeltaLogEntry to UTF-8 YAML bytes.
    /// </summary>
    public byte[] Encode(DeltaLogEntry<TKey, ZSet<TKey>> entry)
    {
        var res = DeltaLogEntryCodec.EncodeYaml(_keyEnc, entry);
        if (res is Result<string, EncodeError>.Ok ok)
        {
            return System.Text.Encoding.UTF8.GetBytes(ok.Value);
        }
        else
        {
            var err = (Result<string, EncodeError>.Err)res;
            throw new InvalidOperationException($"YamlEntryCodec.Encode failed: {err.Error}");
        }
    }

    /// <summary>
    /// Decodes a DeltaLogEntry from UTF-8 YAML bytes.
    /// </summary>
    public DeltaLogEntry<TKey, ZSet<TKey>> Decode(byte[] bytes)
    {
        var yaml = System.Text.Encoding.UTF8.GetString(bytes);
        return DeltaLogEntryCodec.DecodeYaml(_keyDec, yaml, _comparer, _collation);
    }
}
